File: src/main/java/io/playground/orderservice/infrastructure/persistence/order/entity/OrderAmountEntity.java
package io.playground.orderservice.infrastructure.persistence.order.entity;

import io.playground.orderservice.domain.order.OrderAmount;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
@Entity
@Getter
@NoArgsConstructor(access = lombok.AccessLevel.PROTECTED)
@AllArgsConstructor
@Builder
@Table(
        name = "order_amounts",
        indexes = {
                @Index(name = "idx_finalAmount", columnList = "finalAmount")
        }
)
public class OrderAmountEntity {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "order_id", nullable = false)
    private OrderEntity order;

    @Column(nullable = false)
    private BigDecimal totalAmount;

    @Column(nullable = false)
    private BigDecimal discountAmount;

    @Column(nullable = false)
    private BigDecimal deliveryFee;

    @Column(nullable = false)
    private BigDecimal finalAmount;

    public static OrderAmountEntity fromDomain(OrderAmount orderAmount,
                                               OrderEntity orderEntity) {
        return OrderAmountEntity.builder()
                .order(orderEntity)
                .totalAmount(orderAmount.getTotalAmount())
                .discountAmount(orderAmount.getDiscountAmount())
                .deliveryFee(orderAmount.getDeliveryFee())
                .finalAmount(orderAmount.getFinalAmount())
                .build();
    }

    public OrderAmount toDomain() {
        return OrderAmount.of(
                this.id,
                this.order.getId(),
                this.totalAmount,
                this.discountAmount,
                this.deliveryFee,
                this.finalAmount
        );
    }
}

File: src/main/kotlin/io/playground/orderservice/infrastructure/persistence/order/entity/OrderAmountEntity.kt
package io.playground.orderservice.infrastructure.persistence.order.entity

import io.playground.orderservice.domain.order.OrderAmount
import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.FetchType
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.JoinColumn
import jakarta.persistence.ManyToOne
import jakarta.persistence.Table
import java.math.BigDecimal

@Entity
@Table(name = "order_amounts")
class OrderAmountEntity(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Long? = null,
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "order_id", nullable = false)
    var order: OrderEntity? = null,
    @Column(nullable = false)
    var totalAmount: BigDecimal = BigDecimal.ZERO,
    @Column(nullable = false)
    var discountAmount: BigDecimal = BigDecimal.ZERO,
    @Column(nullable = false)
    var deliveryFee: BigDecimal = BigDecimal.ZERO,
    @Column(nullable = false)
    var finalAmount: BigDecimal = BigDecimal.ZERO,
) {
    fun toDomain(): OrderAmount = OrderAmount.of(
        id,
        order?.id,
        totalAmount,
        discountAmount,
        deliveryFee,
        finalAmount,
    )

    companion object {
        @JvmStatic
        fun fromDomain(orderAmount: OrderAmount, orderEntity: OrderEntity): OrderAmountEntity =
            OrderAmountEntity(
                order = orderEntity,
                totalAmount = orderAmount.totalAmount,
                discountAmount = orderAmount.discountAmount,
                deliveryFee = orderAmount.deliveryFee,
                finalAmount = orderAmount.finalAmount,
            )
    }
}
